api/src/test/java/org/apache/iceberg/expressions/TestMetricsEvaluatorsNaNHandling.java

```diff
@@ -0,0 +1,405 @@
+
+    shouldRead = new InclusiveMetricsEvaluator(
+        SCHEMA, Expressions.in("some_nan_correct_bounds", 1F, 10F, 30F)).eval(FILE);
+    Assert.assertTrue("Should match: 10 within bounds", shouldRead);
+
+    shouldRead = new InclusiveMetricsEvaluator(
+        SCHEMA, Expressions.in("some_nan_correct_bounds", 1F)).eval(FILE);
+    Assert.assertFalse("Should not match: 1 not within bounds", shouldRead);
+
+    shouldRead = new InclusiveMetricsEvaluator(
+        SCHEMA, Expressions.in("some_nan_correct_bounds", 30F)).eval(FILE);
+    Assert.assertFalse("Should not match: 30 not within bounds", shouldRead);
+  }
+
+  @Test
+  public void testInclusiveMetricsEvaluatorNotIn() {
+    boolean shouldRead = new InclusiveMetricsEvaluator(
+        SCHEMA, Expressions.notIn("all_nan", 1D)).eval(FILE);
+    shouldRead = shouldRead & new InclusiveMetricsEvaluator(
+        SCHEMA, Expressions.notIn("max_nan", 1D)).eval(FILE);
+    shouldRead = shouldRead & new InclusiveMetricsEvaluator(
+        SCHEMA, Expressions.notIn("max_nan", 10D)).eval(FILE);
+    shouldRead = shouldRead & new InclusiveMetricsEvaluator(
+        SCHEMA, Expressions.notIn("min_max_nan", 1F)).eval(FILE);
+    shouldRead = shouldRead & new InclusiveMetricsEvaluator(
+        SCHEMA, Expressions.notIn("all_nan_null_bounds", 1D)).eval(FILE);
+    shouldRead = shouldRead & new InclusiveMetricsEvaluator(
+        SCHEMA, Expressions.notIn("some_nan_correct_bounds", 1F)).eval(FILE);
+    shouldRead = shouldRead & new InclusiveMetricsEvaluator(
+        SCHEMA, Expressions.notIn("some_nan_correct_bounds", 10F)).eval(FILE);
+    shouldRead = shouldRead & new InclusiveMetricsEvaluator(
+        SCHEMA, Expressions.notIn("some_nan_correct_bounds", 30)).eval(FILE);
+    Assert.assertTrue("Should match: no visibility", shouldRead);
+  }
+
+  @Test
+  public void testStrictMetricsEvaluatorLessThanAndLessThanOrEqual() {
+    for (BiFunction<String, Number, Expression> func : LESS_THAN_EXPRESSIONS) {
+      boolean shouldRead = new StrictMetricsEvaluator(SCHEMA, func.apply("all_nan", 1D)).eval(FILE);
+      Assert.assertFalse("Should not match: all nan column doesn't contain number", shouldRead);
+
+      shouldRead = new StrictMetricsEvaluator(SCHEMA, func.apply("max_nan", 10D)).eval(FILE);
+      Assert.assertFalse("Should not match: 10 is less than upper bound", shouldRead);
+
+      shouldRead = new StrictMetricsEvaluator(SCHEMA, func.apply("min_max_nan", 1F)).eval(FILE);
+      Assert.assertFalse("Should not match: no visibility", shouldRead);
+
+      shouldRead = new StrictMetricsEvaluator(SCHEMA, func.apply("all_nan_null_bounds", 1D)).eval(FILE);
+      Assert.assertFalse("Should not match: all nan column doesn't contain number", shouldRead);
+
+      shouldRead = new StrictMetricsEvaluator(SCHEMA, func.apply("some_nan_correct_bounds", 10F)).eval(FILE);
+      Assert.assertFalse("Should not match: nan value exists", shouldRead);
```

Comment:
This case would return `false` anyway because the correct upper bound is > 10. I think you should change this test case to 30F so that it matches the whole range, but does not match because there are some NaN values.